### ISSUE TITLE
**SendMax** feature is inactive.

### ISSUE THREAD

**Issue by dmigwi** [crypto-power/cryptopower#549]
`SendMax` feature also seems inactive too.

_Originally posted by @dmigwi in https://github.com/crypto-power/cryptopower/issues/548#issuecomment-2184027304_
            

<img width="713" alt="Screenshot 2024-06-22 at 19 25 28" src="https://github.com/crypto-power/cryptopower/assets/22055953/d373411e-2703-4b49-a1b8-45e0d97ea0e7">

**Comment by dmigwi** [comment]
After evaluating the current `SendMax` implementation, I have identified that we have a builtin flaw in how the feature works.

**Problems**
- Clicking the Send Man Button doesn't fill the Send Amount fields with the maximum amount that can be sent. - This means that even after you active the Send max bool, you will need to input an amount to activate the Next button.
- When constructing the a raw transaction, we send both `SendMax` bool and the actual amount to the backend. - This means the functionality can allow a user to unwillingly max out their whole spendable account balance since `SendMax` bool is given higher priority than the user set amount. 

**Solution**
- Move all the send max amounts computation to the frontend. This will prevent duplication of the same Amounts calculations by the backend and frontend.
- Once the tx processing gets to the backend, total amount to be spent could be reconstructed to establish if the user intended to send max amount in the account or not.

**Caveat**
- These changes involve a very critical part of the codebase where comprehensive reviews and testing would need to be done before the changes are accepted.

**Comment by dreacot** [comment]
> When constructing the a raw transaction, we send both SendMax bool and the actual amount to the backend. - This means the functionality can allow a user to unwillingly max out their whole spendable account balance since SendMax bool is given higher priority than the user set amount.

clicking send max should fill the input field with the maximum available spendable balance for the account, and manually modifying the input sets the max button to it's inactive state. This was how it worked before something broke

**Comment by dreacot** [comment]
> These changes involve a very critical part of the codebase where comprehensive reviews and testing would need to be done before the changes are accepted.

i don't think so, we should simply just enable the "max" button when clicked and set the input field to the maximum available spendable balance on the account, and if the input field is manually modified we set the "max" button to it's inactive state

**Comment by ademuanthony** [comment]
> > These changes involve a very critical part of the codebase where comprehensive reviews and testing would need to be done before the changes are accepted.
> 
> i don't think so, we should simply just enable the "max" button when clicked and set the input field to the maximum available spendable balance on the account, and if the input field is manually modified we set the "max" button to it's inactive state

With this, there shouldn’t be a need to send the `SendMax` to the backend; just the value of the input field

**Comment by dreacot** [comment]
> > > These changes involve a very critical part of the codebase where comprehensive reviews and testing would need to be done before the changes are accepted.
> > 
> > 
> > i don't think so, we should simply just enable the "max" button when clicked and set the input field to the maximum available spendable balance on the account, and if the input field is manually modified we set the "max" button to it's inactive state
> 
> With this, there shouldn’t be a need to send the `SendMax` to the backend; just the value of the input field

well the thing is the backend would have the most recent data just right before the tx is created, so notifying the backend the send max is active allows it to make infinitesimal adjustments right before the tx is created

**Comment by dmigwi** [comment]
>clicking send max should fill the input field with the maximum available spendable balance for the account,...

This is not what actually happens. The amount a user inputs is usually the actual amount to be sent excluding the fee to be charged on that tx. i.e. **User Amount input + Tx Fees = Total Cost to be incurred**. This mean that **Max Amount to send = Account Spendable Amount - Tx Fee**. If `SendMax` is false this is not a huge problem otherwise bigger issues arise as highlighted below:
1. [Frontend Tx Construct](https://github.com/crypto-power/cryptopower/blob/c1b583d926015aba8e456b9001153c2d274a1353/ui/page/send/page.go#L343-L398) Implementation doesn't make accurate tx fee calculations compared to the [Backend Tx Construct](https://github.com/crypto-power/cryptopower/blob/c1b583d926015aba8e456b9001153c2d274a1353/libwallet/assets/btc/txauthor.go#L388-L471) Implementation. This means that if the tx gets successfully submitted, the amount the user thought they sent might not be the exact sent by the backend. This possible because `SendMax` and `Amount` variables are set when adding a destination: `AddSendDestination(id int, address string, unitAmount int64, sendMax bool) error`.

2. To fix the problems highlighted in point 1 above, we have to merge `Amount` and `SendMax` variables into a single variable that can pass either of the two informations but not both of them simultaneously. The most complicated part is synchronising the Frontend and Backend Tx constructing functions to have the same output in terms of the Exact Amount to be sent and fees levied on the unsigned tx. Mind you, backend Tx construct function only runs at the tail end of a tx submission while Frontend Tx construct function runs on every minor change detected in the tx input.

>These changes involve a very critical part of the codebase where comprehensive reviews...

3. Resolving the problems identified above will fix the `SendMax` functionality and also display a more accurate Tx cost but this will come at a cost because they will require a significant code overhaul especially in how unsigned txs are constructed. Also during a normal Tx (when `SendMax` is false) a `changeTx` is required but it not considered when computing the txFee for the rawTx. This ought be fixed in the process too.

**Comment by dmigwi** [comment]
In summary:

When `SendMax` is true all `Spendable Amount` should be equal to the `Total Tx Cost`. Since the Frontend unsigned Tx construct doesn't get accurate tx fees, setting the actual amount to be sent on the user input amount field becomes a challenge without a massive code overhaul.

NB: **User Amount input + Tx Fees = Total Cost to be incurred**

**Comment by dreacot** [comment]
the change in amount on the front end and back end has always been that way due to the backend having the most accurate data, that's what i was trying to explain in my other comment https://github.com/crypto-power/cryptopower/issues/549#issuecomment-2247590793, the amount change is usually very negligible to be considered a problem.

having said that, it would be a very good approach to rewrite the logic behind this, but not intended for this issue.

i had hinted to the send max not working properly in previous issues such as this https://github.com/crypto-power/cryptopower/issues/483

right now, we need to get the send max button working again as it has been right from the inception.

which is to enable the "max" button when clicked and set the input field to the maximum available spendable balance on the account, and if the input field is manually modified we set the "max" button to it's inactive state. as long and send max is set to true, the back end handles the necessary logic.

then you can create a new issue, suggesting rewriting the old logic.

**Comment by dmigwi** [comment]
>please make reference to https://github.com/planetdecred/godcr if needed

I have already made reference to that and the only change that seems to have been made is the definition of the user input amount.
- Before: **User Input Amount** == *Total Tx Cost (Send Amount + Tx Fee)*
- Now: **User Input Amount** == *Send Amount (Total Tx Cost - Tx Fee)*

The only two solutions I see is;
-  Revert back to when **User Input Amount** was equal to **Total Tx Cost** - (on the tx construction)
-  Attempt to make more accurate Tx Fee computations.

If we revert back to **User Input Amount** == **Total Tx Cost** a lot of complications might be solved. Also the actual tx fee to be spent is displayed just before a user inputs the account password to sign the tx. This will prevent the use of the inaccurate tx fees during the tx construction stage.

**Comment by dreacot** [comment]
Does the max button currently work for you?

I’m asking based on your "before" and "now" comment
